File: java/src/test/java/it/giannibombelli/wsc2026/booking/application/integration/payment/handlers/HandlePaymentResultFromPaymentTest.java
package it.giannibombelli.wsc2026.booking.application.integration.payment.handlers;

import it.giannibombelli.wsc2026.booking.application.integration.payment.adapter.PaymentResult;
import it.giannibombelli.wsc2026.booking.application.usecases.BookingConfirming;
import it.giannibombelli.wsc2026.booking.application.usecases.BookingRejecting;
import it.giannibombelli.wsc2026.booking.domain.booking.Booking;
import it.giannibombelli.wsc2026.booking.domain.booking.BookingStatus;
import it.giannibombelli.wsc2026.booking.domain.ports.BookingRepository;
import it.giannibombelli.wsc2026.booking.infrastructure.InMemoryBookingEventBus;
import it.giannibombelli.wsc2026.booking.infrastructure.SqliteBookingRepository;
import it.giannibombelli.wsc2026.common.domain.primitive.Money;
import it.giannibombelli.wsc2026.payment.integration.PaymentResultIntegrationEvent;
import it.giannibombelli.wsc2026.testsupport.AggregateFactory;
import it.giannibombelli.wsc2026.testsupport.DatabaseSetup;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

import javax.sql.DataSource;
import java.math.BigDecimal;
import java.util.UUID;

import it.giannibombelli.wsc2026.common.errors.DependencyNotProvidedException;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

class HandlePaymentResultFromPaymentTest {

    private BookingRepository repository;
    private HandlePaymentResultFromPayment handler;

    @BeforeEach
    void setUp() {
        DataSource dataSource = DatabaseSetup.initializeInMemoryDb("booking");
        repository = new SqliteBookingRepository(dataSource);
        InMemoryBookingEventBus eventBus = new InMemoryBookingEventBus(Runnable::run);
        PaymentResult paymentResult = new PaymentResult(repository);
        BookingConfirming confirming = new BookingConfirming(repository, eventBus);
        BookingRejecting rejecting = new BookingRejecting(repository, eventBus);
        handler = new HandlePaymentResultFromPayment(paymentResult, confirming, rejecting);
    }

    @Nested
    class Construction {
        @Test
        void rejectNullParameters() {
            PaymentResult paymentResult = new PaymentResult(repository);
            BookingConfirming confirming = new BookingConfirming(repository, new InMemoryBookingEventBus(Runnable::run));
            BookingRejecting rejecting = new BookingRejecting(repository, new InMemoryBookingEventBus(Runnable::run));

            assertThatThrownBy(() -> new HandlePaymentResultFromPayment(null, confirming, rejecting))
                .isInstanceOf(DependencyNotProvidedException.class);
            assertThatThrownBy(() -> new HandlePaymentResultFromPayment(paymentResult, null, rejecting))
                .isInstanceOf(DependencyNotProvidedException.class);
            assertThatThrownBy(() -> new HandlePaymentResultFromPayment(paymentResult, confirming, null))
                .isInstanceOf(DependencyNotProvidedException.class);
        }
    }

    @Nested
    class PaymentResultsHandling {
        @Test
        void onAccepted_shouldConfirmBooking() {
            Booking booking = saveBooking();
            Money amount = new Money(new BigDecimal("50.00"));

            handler.handle(new PaymentResultIntegrationEvent.PaymentAcceptedIntegrationEvent(
                booking.id().value().toString(), amount
            ));

            Booking updated = repository.findById(booking.id()).orElseThrow();
            assertThat(updated.status()).isEqualTo(BookingStatus.CONFIRMED);
        }

        @Test
        void onRejected_shouldRejectBooking() {
            Booking booking = saveBooking();
            Money amount = new Money(new BigDecimal("50.00"));

            handler.handle(new PaymentResultIntegrationEvent.PaymentRejectedIntegrationEvent(
                booking.id().value().toString(), amount, "declined"
            ));

            Booking updated = repository.findById(booking.id()).orElseThrow();
            assertThat(updated.status()).isEqualTo(BookingStatus.REJECTED);
        }

        @Test
        void onExpired_shouldDoNothing() {
            Booking booking = saveBooking();

            handler.handle(new PaymentResultIntegrationEvent.PaymentExpiredIntegrationEvent(
                booking.id().value().toString(), new Money(new BigDecimal("50.00"))
            ));

            Booking updated = repository.findById(booking.id()).orElseThrow();
            assertThat(updated.status()).isEqualTo(BookingStatus.PLACED);
        }

        @Test
        void onAcceptedUnknownBooking_shouldDoNothing() {
            Money amount = new Money(new BigDecimal("50.00"));

            handler.handle(new PaymentResultIntegrationEvent.PaymentAcceptedIntegrationEvent(
                UUID.randomUUID().toString(), amount
            ));

            assertThat(true).isTrue();
        }
    }

    private Booking saveBooking() {
        Booking booking = AggregateFactory.createBooking();
        repository.save(booking);
        return booking;
    }
}

File: typescript/tests/booking/application/integration/payment/handlers/handlePaymentResultFromPayment.test.ts
import { beforeAll, describe, expect, it } from 'vitest';
import { Money } from '@/common/domain/primitive/money.js';
import {
  paymentAcceptedIntegrationEvent,
  paymentExpiredIntegrationEvent,
  paymentRejectedIntegrationEvent,
} from '@/payment/integration/paymentResultIntegrationEvent.js';
import { HandlePaymentResultFromPayment } from '@/booking/application/integration/payment/handlers/handlePaymentResultFromPayment.js';
import { PaymentResult } from '@/booking/application/integration/payment/adapter/paymentResult.js';
import { BookingConfirming } from '@/booking/application/usecases/bookingConfirming.js';
import { BookingRejecting } from '@/booking/application/usecases/bookingRejecting.js';
import { SqliteBookingRepository } from '@/booking/infrastructure/sqliteBookingRepository.js';
import { InMemoryBookingEventBus } from '@/booking/infrastructure/inMemoryBookingEventBus.js';
import { DatabaseSetup } from '../../../../../testsupport/databaseSetup.js';
import { BookingAggregateFactory } from '../../../../../testsupport/booking/aggregateFactory.js';
import { Uuid } from '@/common/domain/primitive/uuid.js';

describe('HandlePaymentResultFromPayment', () => {
  let repository: SqliteBookingRepository;
  let handler: HandlePaymentResultFromPayment;

  beforeAll(() => {
    const database = DatabaseSetup.initializeInMemoryDb('booking');
    repository = new SqliteBookingRepository(database);
    const eventBus = new InMemoryBookingEventBus((task) => task());
    const paymentResult = new PaymentResult(repository);
    const confirming = new BookingConfirming(repository, eventBus);
    const rejecting = new BookingRejecting(repository, eventBus);
    handler = new HandlePaymentResultFromPayment(paymentResult, confirming, rejecting);
  });

  describe('construction', () => {
    it('rejects null parameters', () => {
      const paymentResult = new PaymentResult(repository);
      const eventBus = new InMemoryBookingEventBus((task) => task());
      const confirming = new BookingConfirming(repository, eventBus);
      const rejecting = new BookingRejecting(repository, eventBus);

      expect(() => new HandlePaymentResultFromPayment(null as unknown as PaymentResult, confirming, rejecting)).toThrow();
      expect(() => new HandlePaymentResultFromPayment(paymentResult, null as unknown as BookingConfirming, rejecting)).toThrow();
      expect(() => new HandlePaymentResultFromPayment(paymentResult, confirming, null as unknown as BookingRejecting)).toThrow();
    });
  });

  describe('payment results handling', () => {
    it('on accepted should confirm booking', () => {
      const booking = BookingAggregateFactory.createBooking();
      repository.save(booking);
      const amount = new Money(50);

      handler.handle(paymentAcceptedIntegrationEvent(booking.id().value.value, amount));

      const updated = repository.findById(booking.id());
      expect(updated).not.toBeNull();
      expect(updated!.status()).toBe('CONFIRMED');
    });

    it('on rejected should reject booking', () => {
      const booking = BookingAggregateFactory.createBooking();
      repository.save(booking);
      const amount = new Money(50);

      handler.handle(paymentRejectedIntegrationEvent(booking.id().value.value, amount, 'declined'));

      const updated = repository.findById(booking.id());
      expect(updated).not.toBeNull();
      expect(updated!.status()).toBe('REJECTED');
    });

    it('on expired should do nothing', () => {
      const booking = BookingAggregateFactory.createBooking();
      repository.save(booking);

      handler.handle(paymentExpiredIntegrationEvent(booking.id().value.value, new Money(50)));

      const updated = repository.findById(booking.id());
      expect(updated).not.toBeNull();
      expect(updated!.status()).toBe('PLACED');
    });

    it('on accepted unknown booking should do nothing', () => {
      const amount = new Money(50);

      handler.handle(paymentAcceptedIntegrationEvent(Uuid.generate().value, amount));

      expect(true).toBe(true);
    });
  });
});
